File: src/main/java/simplifying/conditional/expressions/Calculator.java
package simplifying.conditional.expressions;

import java.util.Map;
import java.util.function.BiFunction;

public class Calculator {
    public static final String ADD = "add";
    public static final String MULTIPLY = "multiply";
    public static final String DIVIDE = "divide";
    public static final String SUBTRACT = "subtract";

    private static final Map<String, BiFunction<Integer, Integer, Integer>> supportedOperations = Map.of(
            ADD, Integer::sum,
            MULTIPLY, (a, b) -> a * b,
            DIVIDE, (a, b) -> a / b,
            SUBTRACT, (a, b) -> a - b
    );

    public static int calculate(int a,
                                int b,
                                String operator) {
        if (!supportedOperations.containsKey(operator)) {
            throw new IllegalArgumentException("Not supported operator");
        }
        return supportedOperations.get(operator).apply(a, b);
    }
}

File: src/scala/src/main/scala/simplifying/conditional/expressions/Calculator.scala
package simplifying.conditional.expressions

import scala.util.{Failure, Success, Try}

object Calculator {
  val add = "add"
  val multiply = "multiply"
  val divide = "divide"
  val subtract = "subtract"

  private val supportedOperators = Map[String, (Int, Int) => Int](
    add -> { (a, b) => a + b },
    multiply -> { (a, b) => a * b },
    divide -> { (a, b) => a / b },
    subtract -> { (a, b) => a - b }
  )

  def calculate(a: Int, b: Int, operator: String): Try[Int] = {
    supportedOperators.get(operator) match {
      case Some(op) => Success(op(a, b))
      case None =>
        Failure(new IllegalArgumentException("Not supported operator"))
    }
  }
}
